### PR TITLE
Master NPA to EVA

### PR DESCRIPTION
**Pull request by max-ruttenberg**
* The branch name is a misnomer. Trust the PR name -- not the branch name.
* Implements the default_npa_to_eva function.
* This should not be merged in until a regression suite has been added to test this feature.

### CODE REVIEW

File: cl_manycore/libraries/bsg_manycore_eva.cpp [hunk 1]
```diff
@@ -354,11 +354,254 @@ int default_eva_to_npa(const hb_mc_config_t *cfg,
...
+ * @param[in] config  An initialized manycore configuration struct
+ * @param[in] npa     An npa to translate
+ * @param[in] tgt     Coordinates of the target tile
+ * @return true if the NPA is DRAM, false otherwise.
+ */
+static bool default_npa_is_dram(const hb_mc_config_t *config,
+                                const hb_mc_npa_t *npa,
+                                const hb_mc_coordinate_t *tgt)
+{
+        return (hb_mc_npa_get_y(npa) == default_get_dram_y(config))
```
Comment (drichmond) [inline cl_manycore/libraries/bsg_manycore_eva.cpp]: I don't think the location of dram should be associated with default - I think it should be associated with the configuration

File: cl_manycore/libraries/bsg_manycore_eva.cpp [hunk 2]
```diff
@@ -354,11 +354,254 @@ int default_eva_to_npa(const hb_mc_config_t *cfg,
...
+                                   const hb_mc_coordinate_t *origin,
+                                   const hb_mc_coordinate_t *tgt,
+                                   const hb_mc_npa_t *npa,
+                                   hb_mc_eva_t *eva,
+                                   size_t *sz)
+{
+        // build the eva
+        hb_mc_eva_t addr = 0;
+        addr |= hb_mc_npa_get_epa(npa); // set the byte address
+        addr |= hb_mc_npa_get_x(npa) << DEFAULT_GLOBAL_X_BITIDX; // set the x coordinate
```
Comment (drichmond) [inline cl_manycore/libraries/bsg_manycore_eva.cpp]: This isn't correct - the X index of DRAM depends on the network address width. DEFAULT_GLOBAL is for accesses to the Local Memory of other cores in the array

Comment (drichmond) [reply cl_manycore/libraries/bsg_manycore_eva.cpp]: See default_eva_to_npa_dram - where shift and mask is computed.

**Comment by max-ruttenberg** [comment]
@drichmond how's it look now?

File: cl_manycore/libraries/bsg_manycore_config.h [hunk 3]
```diff
@@ -155,6 +155,23 @@ static inline hb_mc_idx_t hb_mc_config_get_vcore_base_x(const hb_mc_config_t *cf
         return HB_MC_CONFIG_VCORE_BASE_X; // TODO: These should be defined in the ROM?
 }
 
+static inline size_t hb_mc_config_get_dram_size(const hb_mc_config_t *cfg)
+{
+        return 4ul * (1ul<<30ul); // 4GB: TODO: should read out of the ROM
```
Comment (drichmond) [inline cl_manycore/libraries/bsg_manycore_config.h]: This is determined by the network address size, I think. Right @tommydcjung?

File: cl_manycore/libraries/bsg_manycore_eva.cpp [hunk 4]
```diff
@@ -228,11 +228,11 @@ static int default_eva_to_npa_global(const hb_mc_config_t *cfg,
 	*npa = hb_mc_epa_to_npa(hb_mc_coordinate(x,y), epa);
 
 	bsg_pr_dbg("%s: Translating EVA 0x%x for tile (x: %d y: %d) to NPA {x: %d y: %d, EPA: 0x%x}. \n",
-		__func__, hb_mc_eva_addr(eva),
```
Comment (drichmond) [inline cl_manycore/libraries/bsg_manycore_eva.cpp]: Urgh, so many whitespace changes

**Comment by bornaehsani** [comment]
Works great with CUDA API as well!

**Comment by max-ruttenberg** [comment]
@drichmond nah, let's just merge it. There's a PR in manycore that this PR relies on for the added regression tests to pass.

**Pull request merged**